common/types/ballot.go

```diff
@@ -0,0 +1,166 @@
+package types
+
+import (
+	"bytes"
+
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/signing"
+)
+
+const (
+	// BallotIDSize in bytes.
+	// FIXME(dshulyak) why do we cast to hash32 when returning bytes?
+	BallotIDSize = Hash32Length
+)
+
+// BallotID is a 20-byte sha256 sum of the serialized ballot used to identify a Ballot.
+type BallotID Hash20
+
+// EmptyBallotID is a canonical empty BallotID.
+var EmptyBallotID = &BallotID{}
+
+// Ballot contains the smesher's signed vote on the mesh history.
+type Ballot struct {
+	// the actual votes on the mesh history
+	InnerBallot
+	// smesher's signature on InnerBallot
+	Signature []byte
+
+	// the following fields are kept private and from being serialized
+	ballotID BallotID
+	// the public key of the smesher used
+	smesherID *signing.PublicKey
+}
+
+// InnerBallot contains all info about a smesher's votes on the mesh history. this structure is
+// serialized and signed to produce the signature in Ballot.
+type InnerBallot struct {
+	// the smesher's ATX in the epoch this ballot is cast.
+	AtxID ATXID
+	// the proof of the smesher's eligibility to vote and propose block content in this epoch.
+	EligibilityProof VotingEligibilityProof
+
+	// a smesher creates votes in the following steps:
+	// - select a Ballot in the past as a base Ballot
+	// - calculate the opinion difference on history between the smesher and the base Ballot
+	// - encode the opinion difference in 3 list:
+	//	 - ForDiff
+	//	   contains blocks we support while the base block did not support (i.e. voted against)
+	//	   for blocks we support in layers later than the base block, we also add them to this list
+	//   - AgainstDiff
+	//     contains blocks we vote against while the base block explicitly supported
+	//	 - NeutralDiff
+	//	   contains blocks we vote neutral while the base block explicitly supported or voted against
+	//
+	// example:
+	// layer | unified content block
+	// -----------------------------------------------------------------------------------------------
+	//   N   | UCB_A (genesis)
+	// -----------------------------------------------------------------------------------------------
+	//  N+1  | UCB_B base:UCB_A, for:[UCB_A], against:[], neutral:[]
+	// -----------------------------------------------------------------------------------------------
+	//  N+2  | UCB_C base:UCB_B, for:[UCB_B], against:[], neutral:[]
+	// -----------------------------------------------------------------------------------------------
+	//  (hare hasn't terminated for N+2)
+	//  N+3  | UCB_D base:UCB_B, for:[UCB_B], against:[], neutral:[UCB_C]
+	// -----------------------------------------------------------------------------------------------
+	//  (hare succeeded for N+2 but failed for N+3)
+	//  N+4  | UCB_E base:UCB_C, for:[UCB_C], against:[UCB_D], neutral:[]
```

Comment:
against will not be included in such case, at least in current implementation. also as i understand the current approach is a general agreement with research per https://github.com/spacemeshos/go-spacemesh/issues/2369

i understood that you prefer to set it explicit, but it is not consistent with every other place

Reply:
oh i was mentally thinking when hare timeout in tortoise's view (no result past zdist+confidence param). you are right. in this case it is an implicit against.